src/main/java/org/tikv/common/failsafe/CircuitBreakerImpl.java

```diff
@@ -0,0 +1,187 @@
+import org.slf4j.LoggerFactory;
+import org.tikv.common.TiConfiguration;
+
+public class CircuitBreakerImpl implements CircuitBreaker {
+  private static final Logger logger = LoggerFactory.getLogger(CircuitBreakerImpl.class);
+
+  private final boolean enable;
+  private final int windowInSeconds;
+  private final int errorThresholdPercentage;
+  private final int requestVolumeThreshold;
+  private final int sleepWindowInSeconds;
+  private final int attemptRequestCount;
+
+  private final AtomicLong circuitOpened = new AtomicLong(-1);
+  private final AtomicReference<Status> status = new AtomicReference<>(Status.CLOSED);
+  private final AtomicLong attemptCount = new AtomicLong(0);
+  private final AtomicLong attemptSuccessCount = new AtomicLong(0);
+
+  private final CircuitBreakerMetrics metrics;
+
+  public CircuitBreakerImpl(TiConfiguration conf) {
+    this(
+        conf.isCircuitBreakEnable(),
+        conf.getCircuitBreakAvailabilityWindowInSeconds(),
+        conf.getCircuitBreakAvailabilityErrorThresholdPercentage(),
+        conf.getCircuitBreakAvailabilityRequestVolumnThreshold(),
+        conf.getCircuitBreakSleepWindowInSeconds(),
+        conf.getCircuitBreakAttemptRequestCount());
+  }
+
+  public CircuitBreakerImpl(
+      boolean enable,
+      int windowInSeconds,
+      int errorThresholdPercentage,
+      int requestVolumeThreshold,
+      int sleepWindowInSeconds,
+      int attemptRequestCount) {
+    this.enable = enable;
+    this.windowInSeconds = windowInSeconds;
+    this.errorThresholdPercentage = errorThresholdPercentage;
+    this.requestVolumeThreshold = requestVolumeThreshold;
+    this.sleepWindowInSeconds = sleepWindowInSeconds;
+    this.attemptRequestCount = attemptRequestCount;
+    this.metrics = new CircuitBreakerMetricsImpl(windowInSeconds);
+    this.metrics.addListener(getMetricsListener());
+  }
+
+  private MetricsListener getMetricsListener() {
+    return hc -> {
+      logger.info("onNext " + hc.toString());
+      // check if we are past the requestVolumeThreshold
+      if (hc.getTotalRequests() < requestVolumeThreshold) {
+        // we are not past the minimum volume threshold for the stat window,
+        // so no change to circuit status.
+        // if it was CLOSED, it stays CLOSED
+        // if it was half-open, we need to wait for some successful command executions
+        // if it was open, we need to wait for sleep window to elapse
+      } else {
+        if (hc.getErrorPercentage() < errorThresholdPercentage) {
+          // we are not past the minimum error threshold for the stat window,
+          // so no change to circuit status.
+          // if it was CLOSED, it stays CLOSED
+          // if it was half-open, we need to wait for some successful command executions
+          // if it was open, we need to wait for sleep window to elapse
+        } else {
+          // our failure rate is too high, we need to set the state to OPEN
+          close2Open();
+        }
+      }
```

Comment:
why not simplifying it to:
```java
// Open circuit breaker because the failure is too high. Do nothing for other
// conditions, stay in the current state no matter what it is.
if hc.getTotalRequests() >= requestVolumeThreshold && hc.getErrorPercentage() >= errorThresholdPercentage {
    close2Open();
}
```

Reply:
The only reason is that I want to add the comments in a suitable location.